Adds video upload

fitnr opened the pull request:
This squashes the changes in #929 down to a single commit, incorporating comments from @Harmon758.

Some of this work is due @jamesandres and @Choko256.

This is currently in draft because I have poor internet for the next week and testing video upload isn't feasible. We've waited five years, so another week should be fine. 😄

Harmon758 commented:
I think it would be preferable to keep the original commits if possible, to keep credit where it's due.

fitnr commented:
How about I squash them all into a single commit with multiple authors?

Harmon758 commented:
Is there a reason it needs to be a single commit?

fitnr commented:
Yes. The main branch has changed so much since this was introduced that rebasing each commit is tiresome.

Harmon758 commented:
Wouldn't it be possible to use the original commits already in your video-upload2 branch and make a merge commit that merges the main branch and resolves any conflicts? I think if you really wanted to, you could even just copy all the code as it is in this commit and paste it as part of the merge commit. Although, it'd be preferable for the additional changes in this commit to be separate from changes for conflict resolution as well. Another reason for this is that it'd be easier to specifically review the changes you made in addition to what's already in #929 and the additional commits you have in video-upload2 than to re-review the entire thing.

Maradonna90 commented on tweepy/api.py:
```diff
@@ -242,6 +262,70 @@ def media_upload(self, filename, *args, **kwargs):
+            nloops = int(fsize / chunk_size) + (1 if fsize % chunk_size > 0 else 0)
+
+            for i in range(nloops):
+                headers, post_data, fp = API._chunk_media('append', filename, self.max_size_chunked, chunk_size=chunk_size, f=fp, media_id=media_info.media_id, segment_index=i, is_direct_message=is_direct_message)
```
`self.max_size_chunked` is never defined. Throws errors when I try to test this PR. Should be set accordingly in api.__init__()

Maradonna90 replied:
```
  File "weekly-pitches.py", line 115, in post_tweet
    media_ids = [api.media_upload(video).media_id_string]
  File "/home/marco/weekly-pitches/venv-weekly/lib/python3.8/site-packages/tweepy/api.py", line 245, in media_upload
    return self.upload_chunked(filename, f=f, *args, **kwargs)
  File "/home/marco/weekly-pitches/venv-weekly/lib/python3.8/site-packages/tweepy/api.py", line 276, in upload_chunked
    headers, post_data, fp = API._chunk_media('init', filename, self.max_size_chunked, form_field='media', f=f, is_direct_message=is_direct_message)
AttributeError: 'API' object has no attribute 'max_size_chunked'
```

Harmon758 replied:
Thanks for testing it out. This is actually an old reference from #929 that hasn't been updated.

Harmon758 commented on tweepy/api.py:
```diff
@@ -242,6 +262,70 @@ def media_upload(self, filename, *args, **kwargs):
+        is_direct_message = kwargs.pop('is_direct_message', False)
+
+        # Initialize upload (Twitter cannot handle videos > 15 MB)
+        headers, post_data, fp = API._chunk_media('init', filename, self.max_size_chunked, form_field='media', f=f, is_direct_message=is_direct_message)
```
```suggestion
        headers, post_data, fp = API._chunk_media('init', filename, MAX_UPLOAD_SIZE_CHUNKED, form_field='media', f=f, is_direct_message=is_direct_message)
```

fitnr replied:
I'm adding a separate constant, `MAX_CHUNKSIZE`. According to the docs, [the maximum chunk size is 5MB](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/api-reference/post-media-upload-append), which the maximum total upload size for GIF and video [remains 15MB](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).

Harmon758 commented on tweepy/api.py:
```diff
@@ -242,6 +262,70 @@ def media_upload(self, filename, *args, **kwargs):
+            nloops = int(fsize / chunk_size) + (1 if fsize % chunk_size > 0 else 0)
+
+            for i in range(nloops):
+                headers, post_data, fp = API._chunk_media('append', filename, self.max_size_chunked, chunk_size=chunk_size, f=fp, media_id=media_info.media_id, segment_index=i, is_direct_message=is_direct_message)
```
```suggestion
                headers, post_data, fp = API._chunk_media('append', filename, MAX_UPLOAD_SIZE_CHUNKED, chunk_size=chunk_size, f=fp, media_id=media_info.media_id, segment_index=i, is_direct_message=is_direct_message)
```

Harmon758 commented on tweepy/api.py:
```diff
@@ -242,6 +262,70 @@ def media_upload(self, filename, *args, **kwargs):
+                )(*args, **kwargs)
+
+            # When all chunks have been sent, we can finalize.
+            headers, post_data, fp = API._chunk_media('finalize', filename, self.max_size_chunked, media_id=media_info.media_id, is_direct_message=is_direct_message)
```
```suggestion
            headers, post_data, fp = API._chunk_media('finalize', filename, MAX_UPLOAD_SIZE_CHUNKED, media_id=media_info.media_id, is_direct_message=is_direct_message)
```

fitnr commented:
Thank you @Harmon758  and @Maradonna90 for your help reviewing. I've incorporated your suggested changes.
In particular:
- new global constants are used for tracking the MIN, MAX and DEFAULT chunk sizes. All of these constants are stored in KiB, and multiplied by 1024 when comparing to bytes.
- Enhancement: the `api.media_upload` method will use chunked upload for images that exceed the standard upload size limit
- Repeated logic around checking file types has been removed/simplified
- I added two test files (gif and mp4) and three methods (and casettes) around the media upload endpoint.

savetz commented:
I am eager for the video upload feature, and respectfully ask that it be added to tweepy as soon as possible. Thank you all for your work on this project.

Maradonna90 commented:
If I upload a bigger video and try to post a tweet with it I can get a 324 errorcode with `not valid video`. This happends because the upload hasn't finished yet.

I checked the twitter API and found a method that checks for the upload progress of media. I wrote a small function in my project to use it.

```
def get_media_upload_status(api, *args, **kwargs):
    """ :reference: https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/api-reference/get-media-upload-status
            :allowed_param:
        """

    return bind_api(
            api=api,
            path='/media/upload.json',
            payload_type='media',
            allowed_param=['command', 'media_id'],
            upload_api=True,
            require_auth=True
    )(*args, **kwargs)
```

potentially the media_upload method should return the media_id when the media_upload finished. The above funciton could be used within a routine.

vegit0 commented:
I'm new to using github. How do I use this video upload version as my tweepy for python? What's the pip command for this specific version to install it? Sorry I'm a newbie

savetz commented:
> I'm new to using github. How do I use this video upload version as my tweepy for python? What's the pip command for this specific version to install it? Sorry I'm a newbie

I'm also a newbie, but this is what I just did to get it installed:

git clone https://github.com/fitnr/tweepy.git
cd tweepy/
git checkout video-upload-3
pip3 uninstall tweepy (if you have an old version installed. We're going to replace it)
python3 setup.py build
sudo python3 setup.py install

savetz commented:
I'm gotten it installed, and have successfully to use it to post the included example video.mp4. When I try to post one of my videos, it is successfully assigned a media_ID, but then with the api.update_status command, I get tweepy.error.TweepError: [{'code': 324, 'message': 'Not valid video'}]

This same video works fine if I upload it by hand in the Twitter web interface. It's a little 17KB MP4, definitely not too big.

I don't know if this is a bug tweepy code or if twitter is just being weird about my MP4s.

savetz commented:
I'm gotten it installed, and have successfully to use it to post the included example video.mp4. When I try to post one of my videos, it is successfully assigned a media_ID, but then with the api.update_status command, I get tweepy.error.TweepError: [{'code': 324, 'message': 'Not valid video'}]

This same video works fine if I upload it by hand in the Twitter web interface. It's a little 17KB MP4, definitely not too big.

I don't know if this is a bug tweepy code or if twitter is just being weird about my MP4s.

Maradonna90 commented:
> I'm gotten it installed, and have successfully to use it to post the included example video.mp4. When I try to post one of my videos, it is successfully assigned a media_ID, but then with the api.update_status command, I get tweepy.error.TweepError: [{'code': 324, 'message': 'Not valid video'}]
> 
> This same video works fine if I upload it by hand in the Twitter web interface. It's a little 17KB MP4, definitely not too big.
> 
> I don't know if this is a bug tweepy code or if twitter is just being weird about my MP4s.

have you tried it with `time.sleep(10)` between the `media_upload` call and the `api.update_status` ?

savetz commented:
> have you tried it with `time.sleep(10)` between the `media_upload` call and the `api.update_status` ?

Thanks for the suggestion. I've tried this now. Same thing.

savetz commented:
I found the solution to my Not Valid Video problem. Seems twitter is very fussy about the video codec. 

The ffmpeg command shown here converted my video to a format twitter is happy with
https://gist.github.com/nikhan/26ddd9c4e99bbf209dd7#gistcomment-3232972

and tweepy uploaded it with no complaint!
-K

fitnr commented:
I've added a commit with @Maradonna90's `get_media_upload_status`, it's much more efficient than `time.sleep(10)`

fitnr commented:
@Harmon758 @Maradonna90 bump

Harmon758 reviewed (commented):
I'm going to push this back to v4.0. This would be the only major feature to be added to v3.10, and with it being the last version to support Python 2.7 (and probably Python 3.5), I think it'd be better to have it be released with v4.0, where there's *a lot* of other new features planned, rather than including it in a version that I'd like to be as mature and stable as possible on release. I'd like to begin development of v4.0 as soon as possible, and if any bugs with v3.10 are found after there's commits to the master branch intended for v4.0, I'd probably have to make a separate v3.10.x branch to backport any fixes to release.

I also think there's still a lot of improvements to be made. For example, I'm still of my initial opinion that `_chunk_media` should be refactored into three separate methods.

Regardless, after v3.10 is released, I'll probably merge this into a new separate video-support branch and make a draft PR to merge that branch into the master branch. That will increase visibility and ease of access for anyone wanting to test the feature out before it's released. Also, that way, I'll be able to fix and improve any remaining issues that I find, without having to go through another (and potentially further) reviews, and of course, at that point, any PRs to that branch would be welcome, if anyone else, @fitnr included, wants to make any additional improvements.

Harmon758 commented:
I've gone ahead and made a branch, https://github.com/tweepy/tweepy/tree/video-upload, and merged this into it.
I've also drafted PR #1486 merging it into the master branch.
I'll be making further improvements in that branch and PR, but as I said before, feel free to PR to that branch if anyone else wants to make additional improvements as well.

@vegit0 @savetz See https://tweepy.readthedocs.io/en/latest/install.html and https://pip.pypa.io/en/stable/reference/pip_install/#git.

LuisMayo commented:
The `[{'code': 324, 'message': 'Not valid video'}]` error keeps happening
Shouldn't `media_upload` wait for it before returning?

Harmon758 commented:
Are you using the video-upload branch / PR #1486? This PR has been superseded by that one.

Regardless, [videos need to meet certain specifications for Twitter's API](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/uploading-media/media-best-practices).

> Shouldn't `media_upload` wait for it before returning?

Wait for what?

LuisMayo commented:
I'm using the "video-upload" branch.

Ok let me explain.
When I'm calling
``` python
 uploaded_media = api.media_upload(output_filename, media_category='TWEET_VIDEO')
```
I expect the function to return when the state of the upload is no longer "pending".
Instead I think it's returning right after the "finalize" call. However, after calling finalize you still have to wait until Twitter has ended processing the video, [as explained in the docs](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/api-reference/post-media-upload-finalize):
> it may also be necessary to use a STATUS command and wait for it to return success before proceeding to Tweet creation.

I feel that the same way the API is handling all the process from start to finalize, it should also wait for STATUS not to be "pending" instead of having to handle that outside this library.

I have currently fixed it on my own code using a while loop and waiting for the proper state like this:
``` python
uploaded_media = api.media_upload(output_filename, media_category='TWEET_VIDEO')
while (uploaded_media.processing_info['state'] == 'pending'):
   time.sleep(uploaded_media.processing_info['check_after_secs'])
   uploaded_media = api.get_media_upload_status(uploaded_media.media_id_string)
api.update_status('@' + tweet.author.screen_name + ' ', in_reply_to_status_id=tweet.id_str, media_ids=[uploaded_media.media_id_string])
```

I hope it's clear now.
Thanks

Harmon758 commented:
Ah, I see. Thanks for the feedback.
I think I'll probably add a kwarg to allow waiting for the async finalize process to finish.
I'll look into it later and let you know in #1486.

Harmon758 commented:
The video-upload branch / pull request #1486 should be complete now.
Any feedback or review would be appreciated.
